src/jdk.compiler/share/classes/com/sun/tools/javac/util/MandatoryWarningHandler.java

```diff
@@ -102,34 +106,54 @@ private enum DeferredDiagnosticKind {
 
     /**
      * Create a handler for mandatory warnings.
+     *
+     * @param log     The log on which to generate any diagnostics
+     * @param source  Associated source file, or null for none
+     * @param verbose Specify whether or not detailed messages about
+     *                individual instances should be given, or whether an aggregate
+     *                message should be generated at the end of the compilation.
+     *                Typically set via  -Xlint:option.
+     * @param enforceMandatory
+     *                True if mandatory warnings and notes are being enforced.
+     */
+    public MandatoryWarningHandler(Log log, Source source, boolean verbose, boolean enforceMandatory) {
+        this(log, source, verbose, enforceMandatory, null);
+    }
+
+    /**
+     * Create a handler for mandatory warnings.
+     *
      * @param log     The log on which to generate any diagnostics
+     * @param source  Associated source file, or null for none
      * @param verbose Specify whether or not detailed messages about
      *                individual instances should be given, or whether an aggregate
      *                message should be generated at the end of the compilation.
      *                Typically set via  -Xlint:option.
      * @param enforceMandatory
      *                True if mandatory warnings and notes are being enforced.
-     * @param prefix  A common prefix for the set of message keys for
-     *                the messages that may be generated.
-     * @param lc      An associated lint category for the warnings, or null if none.
+     * @param prefix  A common prefix for the set of message keys for the messages
+     *                that may be generated, or null to infer from the lint category.
      */
-    public MandatoryWarningHandler(Log log, Source source, boolean verbose,
-                                   boolean enforceMandatory, String prefix,
-                                   LintCategory lc) {
+    public MandatoryWarningHandler(Log log, Source source, boolean verbose, boolean enforceMandatory, String prefix) {
         this.log = log;
         this.source = source;
         this.verbose = verbose;
         this.prefix = prefix;
         this.enforceMandatory = enforceMandatory;
-        this.lintCategory = lc;
     }
 
     /**
      * Report a mandatory warning.
+     *
+     * @param pos source code position
+     * @param warnKey lint warning
      */
     public void report(DiagnosticPosition pos, LintWarning warnKey) {
         JavaFileObject currentSource = log.currentSourceFile();
-        Assert.check(warnKey.getLintCategory() == lintCategory);
+
+        // Infer the log prefix from the lint category if not given explicitly
+        if (prefix == null)
```

Comment:
Honestly, I'm not sure if I would call this an improvement. The code used to be strict -- only one kind of warning allowed per handler. Now the handler starts off empty, and the first `report` determines the kind. But the same kind is then used for _all_ deferred logging... my feeling is that this would be better left alone (in the sense that I'm not sure the outcome of this PR is _objectively_ better than the code it replaces?)

Reply:
(of course the part that removes the sunapi handler is ok and orthogonal)

Reply:
I agree, less magic is better :) Thanks for your review - very helpful as always.